File: leetcode/C#/No200.number-of-islands.cs
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * Given a 2d grid map of '1's (land) and '0's (water), count the number of islands.
 * An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically.
 * You may assume all four edges of the grid are all surrounded by water.
 *
 * Example 1:
 * Input:
 * 11110
 * 11010
 * 11000
 * 00000
 * Output: 1
 *
 * Example 2:
 * Input:
 * 11000
 * 11000
 * 00100
 * 00011
 * Output: 3
*/

public class Solution {
    public int NumIslands(char[][] grid) {
        int res = 0;
        int M = grid.Length;
        if (M == 0) return res;
        int N = grid[0].Length;

        void dfs(int i, int j) {
            grid[i][j] = '0';
            foreach (int[] coordinate in new int[][]{
                new int[]{ i + 1, j },
                new int[]{ i - 1, j },
                new int[]{ i, j + 1 },
                new int[]{ i, j - 1 }
            }) {
                if (coordinate[0] < 0 || coordinate[1] < 0 || coordinate[0] >= M || coordinate[1] >= N) continue;
                if (grid[coordinate[0]][coordinate[1]] != '1') continue;
                dfs(coordinate[0], coordinate[1]);
            }
        }

        for (int i = 0; i < M; i += 1) {
            for (int j = 0; j < N; j += 1) {
                if (grid[i][j] == '1') {
                    res += 1;
                    dfs(i, j);
                }
            }
        }
        return res;
    }
}

File: leetcode/Swift/No200.number-of-islands.swift
/*
https://leetcode.com/problems/number-of-islands/
Difficulty:
 Medium
Desc:
Given a 2d grid map of '1's (land) and '0's (water), count the number of islands. An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.
Example:
 Input: grid = [
   ["1","1","1","1","0"],
   ["1","1","0","1","0"],
   ["1","1","0","0","0"],
   ["0","0","0","0","0"]
 ]
 Output: 1
 
 Exaplanation: Whenever you encounter a land, sink all the possible lands you encounter around you and increment the counter by 1
*/
import UIKit
class Medium_099_Number_Of_Islands{
func numIslands(_ grid: [[Character]]) -> Int {
    guard grid.count > 0 else { return 0 }
    guard grid[0].count > 0 else { return 0 }
    var grid = grid
    var islandCounter = 0
    for i in 0..<grid.count{
        for j in 0..<grid[i].count{
            if grid[i][j] == "1"{
                islandCounter += 1
                LandSink(&grid, i, j)
            }
        }
    }
    return islandCounter
}
func LandSink(_ grid : inout [[Character]], _ i : Int, _ j : Int){
    if i >= 0 && j >= 0 && i < grid.count && j < grid[i].count && grid[i][j] == "1" {
        grid[i][j] = "0"
        LandSink(&grid, i+1, j)
        LandSink(&grid, i-1, j)
        LandSink(&grid, i, j+1)
        LandSink(&grid, i, j-1)
    }
    else{
        return
    }
}
}
